---
title: "Progress Report"
author: "Charles Bond"
date: "11/15/2020"
output: pdf_document
bibliography: refsA.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
library(vegan)
```

## Current Status: Stuck, code runs partially, makes it through data setup. Then gives error message, gets stuck on PCA. 

Looking for the tools to run the PCA analysis shown in figure 2 [@RN44], I have been unable to run the code so far. I appear to have all the files, code, and data from the two githubs and the data repository provided in the **Availability of Materials and Data** section of the article of interest. The code looks good, but I can't get past this setup error: the script seems to be having trouble accessing the files that came from the repository.

The very first lines of code, which give me an error, in the Rscript file for the PCA analysis are as follows:

```{r error = TRUE}
## This R-script 
## 
## Author: matt.horton
###############################################################################

rm(list=ls());
require(vegan); require(lme4);
source(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "./hdr.microbiome_methods.R"));
source(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "./hdr.base_methods.R"));



###############################################################################
## metagenomic variables:
###############################################################################
numberOfPCsMicrobes <- 5;
organ <- "root";
phylogeneticMarker <- "ITS";
otuCutoff <- 97;
minimumReads <- 1;
otuThreshold <- 2;

normalizationMethod <- "raw";
analyticalMethod <- "glmer";
ordinationMethod <- "pca";
###############################################################################
## end of metagenomic variables. 
###############################################################################

{
	stopifnot( normalizationMethod != "hlgr" ); ## double-negative
	possibleAnalyticalMethods <- c("lmer", "glmer", "logit" );
	analyticalMethod <- match.arg( analyticalMethod, choices=possibleAnalyticalMethods);
	ordinationMethod <- match.arg( ordinationMethod, c("pca", "cca", "dca"));

	if( normalizationMethod != "pa" & analyticalMethod == "logit" ){
		cat("Changing normalization method for ", phylogeneticMarker, ".\n");
		normalizationMethod <- "pa";
		stopifnot(minimumReads == 250);

	} else if( normalizationMethod != "raw" & analyticalMethod == "glmer" ){
		normalizationMethod <- "raw";
		cat("Changing normalization method for ", phylogeneticMarker, " to: ", normalizationMethod, "\n");
		stopifnot(minimumReads == 1);
	}

	###############################################################################
	## OPEN the dataset
	###############################################################################
	listOfData <- getData(
			cutoff=otuCutoff,
			combined=FALSE,
			marker=phylogeneticMarker,
			tissue=organ, 
			minReadsPerSite=minimumReads, 
			minSamplePerOTU=1, 
			whetherToGetTaxa=TRUE, 
			rarefy=(minimumReads > 1),
			numberOfRarefactions=1);

	speciesMatrix <- listOfData$data;
	filenamePrefix <- basename(listOfData$filename);

	###############################################################################
	## estimate the effort/offset
	###############################################################################
	data.offsets <- colSums( speciesMatrix );
	kingdom <- ifelse( phylogeneticMarker == "16S", "bacteria", "fungi" );
	cat("Analyzing:", nrow(speciesMatrix), "OTUs, surveyed in:", ncol(speciesMatrix), "samples\n");

	transformation <- ordinationMethod; # begin writing the suffix.

	###############################################################################
	## eliminate rare OTUs with a threshold...
	## z.B. doing this with the threshold == 2 eliminates OTUs that occur in only one sample (singletons).
	###############################################################################
	if( !is.na( normalizationMethod )){
		speciesMatrix <- normalizeInSomeWay( speciesMatrix, normalizationMethod = normalizationMethod, eliminateSingletons=(otuThreshold > 1));
	}

	filenamePrefix <- gsub("1sampPerOTU\\.", "", filenamePrefix);

	###############################################################################
	##
	###############################################################################
	numbersToTest <- sort(c(25, 50, 100 )); #ceiling(speciesCutoffs * nrow(speciesMatrix))));
	speciesMatrix <- speciesMatrix[order(rowSums(speciesMatrix), decreasing=T),];
	speciesMatrix <- speciesMatrix[1:max(numbersToTest),];

	###############################################################################	
	################################## GET BLUPS ##################################
	###############################################################################	
	{
		cat("------------------------------\n");
		speciesMatrix <- getBlups( kingdom, analyticalMethod, speciesMatrix, data.offsets );

		## if lots of tests fail, we have to reduce this variable to account for the shrinkage!!!
		if(ncol(speciesMatrix) < max(numbersToTest)){ 
			numbersToTest[which(numbersToTest > ncol(speciesMatrix))] <- ncol(speciesMatrix); ## in case some of the models didn't converge.
			numbersToTest <- unique(numbersToTest);
		}
	}

	###############################################################################	
	## Move to the output directory.
	###############################################################################
	setwd(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "/gwas/otus", otuCutoff, "/", organ, "/phenotypes/")); ## 
	cat("Performing", ordinationMethod, "\n");

	formatEigenvectorsForGWA <- function( eigenvectorMatrix, numberOfEigenvectors=numberOfPCs){
		outputMatrix <- data.frame(phenotype_id=numeric(), phenotype_name=character(), ecotype_id=numeric(), value=numeric(), replicate_id=numeric());

		for( j in 1:numberOfEigenvectors ){
			phenotype_j <- mstack(eigenvectorMatrix[,j], newHeaders=c("ecotype_id", "value"), sorted=F);

			eig_j <- data.frame(
					phenotype_id=rep(j, nrow(phenotype_j)), 
					phenotype_name=rep(colnames(eigenvectorMatrix)[j], nrow(phenotype_j)), 
					ecotype_id=phenotype_j[,"ecotype_id"], value=phenotype_j[,"value"], replicate_id=rep(1, nrow(phenotype_j)), stringsAsFactors=FALSE);

			outputMatrix <- rbind(outputMatrix, eig_j);
		}

		return(outputMatrix);
	}

	###############################################################################	
	## now that things are in order: let us ordinate and write out the phenotype file (mixmogam file format)
	###############################################################################	
	for( j in 1:length(numbersToTest)){
		nspecies_j <- numbersToTest[j];
		cat("Conducting", ordinationMethod, "using the top:", nspecies_j, "'species' in the", kingdom, "\n");
		subset_j <- speciesMatrix[,1:nspecies_j];
		suffix <- paste0(analyticalMethod, ".", ordinationMethod, ".n", nspecies_j);

		if( ordinationMethod == "pca" ){
			pca.obj <- prcomp( subset_j, scale=TRUE, center=TRUE);
			u <- pca.obj$x[,1:numberOfPCsMicrobes];

		} else {
			stop(paste0( ordinationMethod, "is not supported yet.\n"));
		}

		colnames(u) <- paste0(filenamePrefix, ".tn", otuThreshold, ".", suffix, "_", colnames(u));

		outputFile <- formatEigenvectorsForGWA(u, numberOfEigenvectors=numberOfPCsMicrobes);
		outputFileName <- paste0(filenamePrefix, ".tn", otuThreshold, ".", suffix, ".txt");
		write.table(outputFile, file=outputFileName, quote=F, sep=",", row.names=F);
	}
}
```

The `source` function cannot access the required file. I checked that the requested files are indeed in the working directory, and I have tried moving them to the main folder and checking the code, I don't know why R can not access these files. 

The README says:

To avoid using hardcoded paths, we use R's ability to load environmental variables (using the command syntax Sys.getenv('variable_name')).
The variables that need to be loaded, by the user, can be done so using the following bash code:
```console
export ROOT_MICROBIOME_DIR='<your_directory>/root_microbes/'
```

I have not yet been able to figure out what to do with that information. 


I wanted to run some kind of PCA even if only a subset of the data, but the PCA is based on metagenoimc data which I have never worked with. Just looking at it I can see that there is a lot of code involved in making the data usable, which I wouldn't know how to do without even the first line of the original code to work. 

```{r combinedkingdoms}

# TODO: Add comment
# 
# Author: matt.horton
###############################################################################

rm(list=ls());

require(vegan); require(lme4);

source(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "./hdr.microbiome_methods.R"));
source(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "./hdr.base_methods.R"));

###############################################################################
## metagenomic variables:
###############################################################################
centerPCA <- scalePCA <- TRUE;
possibleAnalyticalMethods <- c("lmer", "glmer", "logit");

numberOfPCsMicrobes <- 5;
organ <- "root";

otuCutoff <- 97;
otuThreshold <- 2;
minReadsBacteria <- minReadsFungi <- 1;

ordinationMethod <- "pca";

normalizationMethod16S <- normalizationMethodITS <- "raw";
analyticalMethodBacteria <- "glmer";
analyticalMethodFungi <- "glmer"; ## logistic regression requires the argument `logit`, which is a binomial family instead of a poisson family model (glmer)

###############################################################################
## end of metagenomic variables:
###############################################################################

{
	analyticalMethodBacteria <- match.arg( analyticalMethodBacteria, choices=possibleAnalyticalMethods);
	if( normalizationMethod16S != "pa" & analyticalMethodBacteria == "logit" ){
		cat("Changing normalization method for bacteria.\n");
		normalizationMethod16S <- "pa";
		stopifnot(minReadsBacteria == 250);
		
	} else if( normalizationMethod16S != "raw" & analyticalMethodBacteria == "glmer" ){
		cat("Changing normalization method for bacteria.\n");
		normalizationMethod16S <- "raw";
		stopifnot(minReadsBacteria == 1);
	}
	
	analyticalMethodFungi <- match.arg( analyticalMethodFungi, choices=possibleAnalyticalMethods);	
	if( normalizationMethodITS != "pa" & analyticalMethodFungi == "logit" ){
		cat("Changing normalization method for fungi.\n");
		normalizationMethodITS <- "pa";
		minReadsFungi <- 250;

	} else if( normalizationMethodITS != "raw" & analyticalMethodFungi == "glmer" ){
		cat("Changing normalization method for fungi.\n");
		normalizationMethodITS <- "raw";
		stopifnot(minReadsBacteria == 1);
	}

	###############################################################################
	## BACTERIAL data
	###############################################################################
	bacteria <- getData(
			cutoff=otuCutoff,
			combined=FALSE,
			marker="16S",
			tissue=organ, 
			minReadsPerSite=minReadsBacteria, 
			minSamplePerOTU=1, 
			whetherToGetTaxa=TRUE, 
			rarefy=(minReadsBacteria > 1),
			numberOfRarefactions=1)$data;

	###############################################################################
	## FUNGAL data
	###############################################################################
	fungi <- getData(
			cutoff=otuCutoff,
			combined=FALSE,
			marker="ITS",
			tissue=organ, 
			minReadsPerSite=minReadsFungi, 
			minSamplePerOTU=1, 
			whetherToGetTaxa=TRUE, 
			rarefy=(minReadsFungi > 1),
			numberOfRarefactions=1)$data;

	## if we support minReads == 1, offsets need to be calculated before normalization
	bacterialOffsets <- colSums(bacteria);
	fungalOffsets <- colSums(fungi);

	###############################################################################
	## find the overlap of the two datasets!
	###############################################################################
	cat("There are:", nrow(bacteria), "bacterial OTUs, surveyed in:", ncol(bacteria), "samples\n");
	cat("There are:", nrow(fungi), "fungal OTUs, surveyed in:", ncol(fungi), "samples\n");
	cat("Now sorting by abundance.\n");

	names <- intersect(colnames(bacteria), colnames(fungi));
	bacteria <- bacteria[, names];
	bacteria <- bacteria[order(rowSums(bacteria), decreasing=T),];
	bacterialOffsets <- bacterialOffsets[names];

	fungi <- fungi[, names];
	fungi <- fungi[order(rowSums(fungi), decreasing=T),];
	fungalOffsets <- fungalOffsets[names];
	stopifnot(all.equal(colnames(bacteria), colnames(fungi)));

	###############################################################################
	## eliminate rare OTUs with a threshold...
	## z.B. doing this with the threshold == 2 eliminates OTUs that occur in only one sample (singletons).
	###############################################################################
	if( !is.na( normalizationMethod16S )){
		bacteria <- normalizeInSomeWay( bacteria, normalizationMethod = normalizationMethod16S, eliminateSingletons=(otuThreshold > 1));	
	}

	if( !is.na( normalizationMethodITS )){
		fungi <- normalizeInSomeWay( fungi, normalizationMethod = normalizationMethodITS, eliminateSingletons=(otuThreshold > 1));	
	}

	###############################################################################
	## get the blups for the bacteria and fungi
	###############################################################################
	numbersToTestBacteria <- 100; #ceiling(speciesCutoffs * nrow(bacteria))));
	bacteria <- bacteria[1:max(numbersToTestBacteria),];

	numbersToTestFungi <- c(25, 100); #ceiling(speciesCutoffs * nrow(fungi))));
	fungi <- fungi[1:max(numbersToTestFungi),];

	formatEigenvectorsForGWA <- function( eigenvectorMatrix, numberOfEigenvectors=numberOfPCs){
		outputMatrix <- data.frame(phenotype_id=numeric(), phenotype_name=character(), ecotype_id=numeric(), value=numeric(), replicate_id=numeric());

		for( j in 1:numberOfEigenvectors ){
			phenotype_j <- mstack(eigenvectorMatrix[,j], newHeaders=c("ecotype_id", "value"), sorted=F);

			eig_j <- data.frame(
					phenotype_id=rep(j, nrow(phenotype_j)), 
					phenotype_name=rep(colnames(eigenvectorMatrix)[j], nrow(phenotype_j)), 
					ecotype_id=phenotype_j[,"ecotype_id"], value=phenotype_j[,"value"], replicate_id=rep(1, nrow(phenotype_j)), stringsAsFactors=FALSE);
			
			outputMatrix <- rbind(outputMatrix, eig_j);
		}

		return(outputMatrix);
	}

	###############################################################################	
	################################## BACTERIA ###################################
	###############################################################################	
	{
		cat("Getting the blups for the bacteria.\n");
		bacteria <- getBlups( "bacteria", analyticalMethodBacteria, bacteria, bacterialOffsets );

		## if tests fail, we may have to reduce this variable to account for the shrinkage!!!
		if(ncol(bacteria) < numbersToTestBacteria[length(numbersToTestBacteria)]){
			numbersToTestBacteria[which(numbersToTestBacteria > ncol(bacteria))] <- ncol(bacteria); ## in case some of the models didn't converge.
			numbersToTestBacteria <- unique(numbersToTestBacteria);
		}
	}

	###############################################################################	
	#################################### FUNGI ####################################
	###############################################################################	
	{
		cat("Getting the blups for the fungi.\n");
		fungi <- getBlups( "fungi", analyticalMethodFungi, fungi, fungalOffsets );

		## if tests fail, we may have to reduce this variable to account for the shrinkage!!!
		if(ncol(fungi) < numbersToTestFungi[length(numbersToTestFungi)]){
			numbersToTestFungi[which(numbersToTestFungi > ncol(fungi))] <- ncol(fungi); ## in case some of the models didn't converge.
			numbersToTestFungi <- unique(numbersToTestFungi);
		}
	}

	{
		################################################################################################################################################
		# specify the mapping directory.
		################################################################################################################################################
		setwd(paste0(Sys.getenv('ROOT_MICROBIOME_DIR'), "/gwas/otus", otuCutoff, "/", organ, "/phenotypes/")); ## 
		minimumReads <- paste( unique(c( minReadsBacteria, minReadsFungi)), sep="", collapse="_"); rm(minReadsBacteria); rm(minReadsFungi);

		###############################################################################
		## combine subsets of the bacterial and fungal communities
		## to conduct GWAS, write out the phenotype file.
		###############################################################################
		cat("------------------------------\n");
		cat("------------------------------\n");
		cat("Performing", ordinationMethod, "\n");
		cat("------------------------------\n");
		cat("------------------------------\n");

		for( j in 1:length( numbersToTestBacteria )){
			cat("------------------------------\n");
			nspecies_bacteria <- numbersToTestBacteria[j];
			bac.subset_j <- bacteria[,1:nspecies_bacteria];			
			cat("Working on bacterial sample size:", nspecies_bacteria, "\n");

			for( k in 1:length( numbersToTestFungi )){
				nspecies_fungi <- numbersToTestFungi[k];
				cat("Working on fungal sample size:", nspecies_fungi, "\n");

				fun.subset_k <- fungi[,1:nspecies_fungi];

				###############################################################################
				## now combine everything and THEN conduct PCA, and write out the phenotype file
				###############################################################################
				## cbind calls data.frame, which doesn't check names by default
				stopifnot(sum(rownames(bac.subset_j) == rownames(fun.subset_k)) == nrow(bac.subset_j));
				compositeCommunity <- cbind(bac.subset_j, fun.subset_k); 

				compositeCommunity.pca <- prcomp(compositeCommunity, center=centerPCA, scale=scalePCA);
				u <- compositeCommunity.pca$x[,1:numberOfPCsMicrobes];

				filenamePrefix <- paste0(organ, ".", minimumReads,  ".combined_", ordinationMethod, ".tn", otuThreshold, ".bac_n", nspecies_bacteria, "_", analyticalMethodBacteria, ".fun_n", nspecies_fungi, "_", analyticalMethodFungi, "");
				colnames(u) <- paste0(filenamePrefix, colnames(u));

				outputFile <- formatEigenvectorsForGWA(u, numberOfEigenvectors=numberOfPCsMicrobes);
				outputFileName <- paste0(filenamePrefix, ".txt");
				write.table(outputFile, file=outputFileName, quote=F, sep=",", row.names=F);
				cat("------------------------------\n");
			}
		}
	}
}



```




# Previous Update, 11/08/2020

Based on the availability of the data and the posting of the code on github ( https://github.com/bvilhjal/mixmogam and https://github.com/mahort/root_microbiome.
 ) I have selected a paper on the *Arabidopsis* root microbiome [@RN44] as the subject of my next replication attempt. If it goes well there are a lot of figures in this paper that would be worthwhile to replicate, but the one I am mmost interested in is the PCA shown in Figure 2 of the paper. 

Once again, this paper uses the package `vegan` to do the PCA, so I've got that.

## Status: Making progress, found PCA file in github and just trying to figure out where to start.
I am confused on how approach this, like I made a github repository for my replication, but they have a huge github repository for the original, and I'm wondering if I should download files individually or try to pull the whole project down in case there are files it's referncing that I miss?
I downloaded whole github repository as a zip file and put it here, but I guess I need to alter relative paths or something to make it work? I tried running one of the PCA Rscript and get the error
"Error in getData(cutoff = otuCutoff, combined = FALSE, marker = phylogeneticMarker,  : 
  unused arguments (cutoff = otuCutoff, combined = FALSE, marker = phylogeneticMarker, tissue = organ, minReadsPerSite = minimumReads, minSamplePerOTU = 1, whetherToGetTaxa = TRUE, rarefy = (minimumReads > 1), numberOfRarefactions = 1)", 
  which idk what that means but I guess I need to tinker with this more. 

## References

